Add support for ESM plugins (#7339)
stylelint/stylelint: docs/developer-guide/plugins.md

## The anatomy of a plugin

Let's say we write a plugin forbidding certain word (e.g., "foo") within a selector.

> The plugin's implementation would be:

Reviewer: This example plugin disallows the word "foo" in selectors:

Author: Thanks. Fixed via 3288290af.

```js
import stylelint from "stylelint";

> const { report, ruleMessages, validateOptions } = stylelint.utils;

Author: Agree. Fixed via 06b94d8b9.

const ruleName = "foo-org/selector-no-foo";

const messages = ruleMessages(ruleName, {
  rejected: (selector) => `Unexpected "foo" within selector "${selector}"`
});

const meta = {
  url: "https://github.com/foo-org/stylelint-selector-no-foo/blob/main/README.md"
};

const ruleFunction = (primary, secondaryOptions) => {
  return (root, result) => {
    const validOptions = validateOptions(result, ruleName, {
      actual: primary,
      possible: [true]
    });

>     if (!validOptions) {
>       return;
>     }

Reviewer: ```suggestion
    if (!validOptions) return;
```

To match our code style.

    root.walkRules((ruleNode) => {
      const { selector } = ruleNode;

      if (!selector.includes("foo")) return;

      report({
        result,
        ruleName,
        message: messages.rejected(selector),
        node: ruleNode,

>         word: selector

Reviewer 2: question

Should `index` and `endIndex` be mentioned somewhere?

Author: Good catch. I've tried adding a new description to the `report()` sections (f83157c92). Any thoughts?

Reviewer 2: > Any thoughts?

> For example:

has no examples.

> _Do not use…

Maybe use an `[!IMPORTANT]` instead?

Author: Thanks. Fixed the typos and changed to `[!IMPORTANT]` (af6d74393)

>       if (selector.includes("foo")) {
>         report({
>           result,
>           ruleName,
>           message: messages.rejected(selector),
>           node: ruleNode
>           word: selector,
>         });
>       }

Reviewer: ```suggestion
      if (!selector.includes("foo")) return;
      
      report({
        result,
        ruleName,
        message: messages.rejected(selector),
        node: ruleNode
        word: selector,
      });
```

minor to advocate for our pattern of returning early in rules.

Author: Fixed bde97188d.

ruleFunction.messages = messages;
ruleFunction.meta = meta;

> export default stylelint.createPlugin(ruleName, ruleFunction);

Reviewer: ```suggestion
export default createPlugin(ruleName, ruleFunction);
```

```

> > [!IMPORTANT]
> > A CommonJS plugin also works, but such a plugin may be unsupported in the future.
> > We recommend writing an ES module plugin if compatibility doesn't matter.

Reviewer: ```suggestion
```

Let's take this out for now, and add something about commonjs plugins being deprecated in a follow-up issue.

### Asynchronous rules

You can write your plugin as an _async function_ to deal with `Promise`.

```js
const ruleFunction = (primary, secondaryOptions) => {
  return async (root, result) => {
    // validate options...

>     // load a forbidden word asynchronously
>     const forbiddenWord = await import("./forbidden-word.js");

Reviewer: ```suggestion
    // load disallowed words asynchronously
    const disallowedWords = await import("./disallowed-words.js");
```

our consistent language

    // traverse AST nodes...

>     // report a warning if a violation against the word is detected...

Reviewer: ```suggestion
    // report a warning if a problem word is detected...
```

## `stylelint.utils`

Stylelint exposes some useful utilities.

You're also welcome to copy any of the [internal utils](https://github.com/stylelint/stylelint/tree/main/lib/utils) into your plugin. You should not `import` them directly, as they are not part of the public API and may change or be removed without warning.

### `stylelint.utils.report()`

Adds problems from your plugin to the list of problems that Stylelint will report to the user.

Use `stylelint.utils.report()` to ensure your plugin respects disabled ranges and other possible future features of stylelint. _Do not use PostCSS's `node.warn()` method directly._

### `stylelint.utils.ruleMessages()`

Tailors your messages to the format of standard Stylelint rules.

### `stylelint.utils.validateOptions()`

Validates the options for your rule.

### `stylelint.utils.checkAgainstRule()`

Checks CSS against a standard or custom Stylelint rule _within your own rule_. This function provides power and flexibility for plugins authors who wish to modify, constrain, or extend the functionality of existing Stylelint rules.

> [!NOTE]

> > This is an async function. Your custom rule may need to wait until a `Promise` the function returns is resolved.

Reviewer 2: > a `Promise` the function returns

There's a typo here.

Author: Good catch. Fixed via b703a8f0e.

- `result?`: the PostCSS result for resolving and invoking custom rules
- `context?`: the [context](rules.md#add-autofix) for the rule you are invoking

Use the warning to create a _new_ warning _from your plugin rule_ that you report with [`stylelint.utils.report`](#stylelintutilsreport).

For example, imagine you want to create a plugin that runs [`at-rule-no-unknown`](../../lib/rules/at-rule-no-unknown/README.md) with a built-in list of exceptions for at-rules provided by your preprocessor-of-choice:

```js

> const { checkAgainstRule, report } = stylelint.utils;

Reviewer: ```suggestion
const { utils: checkAgainstRule, report } = stylelint;
```

const allowableAtRules = [
  /* .. */
];

const ruleName = "your-own/at-rule-no-unknown";

const myPluginRule = (primary, secondaryOptions, context) => {
  return async (root, result) => {
    const ignoreAtRules = allowableAtRules.concat(
      secondaryOptions?.ignoreAtRules ?? []
    );
    const defaultedSecondaryOptions = { ...secondaryOptions, ignoreAtRules };

    await checkAgainstRule(
      {
        ruleName: "at-rule-no-unknown",
        ruleSettings: [primary, defaultedSecondaryOptions],
        root,
        result,
        context
      },
      (warning) => {
        report({
          ruleName,
          result,
          message: warning.text,
          node: warning.node,
          start: { line: warning.line, column: warning.column },
          end: { line: warning.endLine, column: warning.endColumn }
        });
      }
    );
  };
};
```

## `stylelint.rules`

All of the rule functions are available at the `stylelint.rules` object. This allows you to build on top of existing rules for your particular needs.

> [!NOTE]

> > Every value in the `stylelint.rules` object is a `Promise` resolving a rule function. This avoids loading all the rules and slowing the start-up down.

Reviewer: ```suggestion
> Every value in the `stylelint.rules` object is a `Promise` resolving a rule function.
```

removed internal implementation rationale.

Here's an example of a plugin that runs `declaration-no-important` only if there is a special directive `@@check-declaration-no-important` somewhere in the stylesheet:

```js

> stylelint.createPlugin(ruleName, (primary) => {

Reviewer: ```suggestion
createPlugin(ruleName, (primary) => {
```
